lib/bitbucket/bitbucket.go

```diff
@@ -0,0 +1,78 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package bitbucket
+
+import (
+	"github.com/gravitational/trace"
+	"github.com/mitchellh/mapstructure"
+)
+
+// IDTokenClaims
+// See the following for the structure:
+// https://support.atlassian.com/bitbucket-cloud/docs/integrate-pipelines-with-resource-servers-using-oidc/
+type IDTokenClaims struct {
+	// Sub provides some information about the Bitbucket Pipelines run that
+	// generated this token. Format: {RepositoryUUID}:{StepUUID}
+	Sub string `json:"sub"`
+
+	// StepUUID is the UUID of the pipeline step for which this token was
+	// issued. Bitbucket UUIDs must begin and end with braces, e.g. '{...}'
+	StepUUID string `json:"stepUuid"`
+
+	// RepositoryUUID is the UUID of the repository for which this token was
+	// issued. Bitbucket UUIDs must begin and end with braces, e.g. '{...}'.
+	// This value may be found in the Pipelines -> OpenID Connect section of the
+	// repository settings.
+	RepositoryUUID string `json:"repositoryUuid"`
+
+	// PipelineUUID is the UUID of the pipeline for which this token was issued.
+	// Bitbucket UUIDs must begin and end with braces, e.g. '{...}'
+	PipelineUUID string `json:"pipelineUuid"`
+
+	// WorkspaceUUID is the UUID of the workspace for which this token was
+	// issued. Bitbucket UUIDs must begin and end with braces, e.g. '{...}'.
+	// This value may be found in the Pipelines -> OpenID Connect section of the
+	// repository settings.
+	WorkspaceUUID string `json:"workspaceUuid"`
+
+	// DeploymentEnvironmentUUID is the name of the deployment environment for
+	// which this pipeline was executed. Bitbucket UUIDs must begin and end with
+	// braces, e.g. '{...}'.
+	DeploymentEnvironmentUUID string `json:"deploymentEnvironmentUuid"`
+
+	// BranchName is the name of the branch on which this pipeline executed.
+	BranchName string `json:"branchName"`
+}
+
+// JoinAuditAttributes returns a series of attributes that can be inserted into
+// audit events related to a specific join.
+func (c *IDTokenClaims) JoinAuditAttributes() (map[string]interface{}, error) {
+	res := map[string]interface{}{}
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
```

Comment:
Why not use encoding/json here?

Reply:
Hm - from what I can recall, using map[string]interface{} allows us to put this into a protobuf audit event as structpb Struct, which is why I went for this. It feels a little odd to hold marshalled json inside of a protobuf.

Reply:
the correct way of handling to structpb.Struct is to marshal it to json either using jsonpb or json in case it's gogo proto, and later unmarshal to the structpb.Struct

Reply:
Yeah, looks like we do that properly when actually encoding into audit events or the database, using `apievents.EncodeMap()`: 
* `join.go`: https://github.com/gravitational/teleport/blob/master/lib/auth/join.go#L146
* `EncodeMap()`: https://github.com/gravitational/teleport/blob/master/api/types/events/struct.go#L101

I didn't fully understand the intent here before, but I think we're just trying to convert a struct to K/V pairs (i.e. `map[string]interface{}`) rather than to a JSON string, which is the internal representation we eventually pass along to `EncodeMap()`.

The struct was originally parsed from a JSON document (the JWT) so it already has tags to map to and from appropriate named fields; `mapstructure` is just a convenient way to do the conversion automatically instead of manually writing out the map.

I don't think `encoding/json` is able to help with that since we aren't going to/from a JSON string here.

Reply:
> mapstructure is just a convenient way to do the conversion automatically instead of manually writing out the map.

Yeah - this maps to what I was thinking about when I originally added this. The goal was to be able to added metadata from the attestation process into the audit log. 90% of our join method are based on some kind of JWT, but we have a few which are not. 

I'm not particularly keen on the implementation I chose - but - if we're going to change this, we should address this in a separate PR/issue and update all of the join methods at once since this is in an interface on the join.go side.